cfme/networks/topology.py

```diff
@@ -0,0 +1,206 @@
+
+            for element in found_elements:
+                self.elements_obj.append(TopologyElement(obj=self, element=element))
+
+            for line in Topology.browser.elements(self.LINES):
+                self.lines_obj.append(TopologyLine(element=line))
+
+    def reload_legends(self):
+        self.legends_obj = []
+        self.display_names = TopologyDisplayNames()
+
+        found_legends = Topology.browser.elements(self.LEGENDS)
+        for legend in found_legends:
+            legend_text = Topology.browser.text(legend.find_element_by_tag_name('label'))
+
+            legend_object = TopologyLegend(name=legend_text, element=legend)
+            self.legends_obj.append(legend_object)
+
+    def movement_stopped(self):
+        element = TopologyElement(obj=self, element=Topology.browser.elements(self.ELEMENTS)[-1])
+        if element.x == self.element_ref.x and element.y == self.element_ref.y:
+            return True
+        self.element_ref = element
+        return False
+
+    @property
+    def legends(self):
+        return self.legends_obj
+
+    @property
+    def elements(self):
+        return self.elements_obj
+
+    @property
+    def lines(self):
+        return self.lines_obj
+
+
+@navigator.register(Topology, 'All')
+class All(CFMENavigateStep):
+    prerequisite = NavigateToAttribute('appliance.server', 'LoggedIn')
+    VIEW = TopologyView
+
+    def step(self):
+        self.prerequisite_view.navigation.select('Networks', 'Topology')
+
+
+class TopologyLegend(object):
+    def __init__(self, name, element):
+        self.legend_name = name
+        self.element = element
+
+    @property
+    def name(self):
+        return self.legend_name
+
+    @property
+    def is_active(self):
+        return 'active' in self.element.get_attribute('class')
+
+    def set_active(self, active=True):
+        if active != self.is_active:
+            self.element.click()
+
+
+class TopologyDisplayNames(object):
+    DISPLAY_NAME = '|'.join([
+        "//*[contains(@class, 'container_topology')]//label[contains(., 'Display Names')]/input",
+        '//*[@id="box_display_names"]'])  # [0] is not working on containers topology
+
+    def __init__(self):
+        self.element = Topology.browser.element(self.DISPLAY_NAME)
+
+    @property
+    def is_enabled(self):
+        return self.element.is_selected()
+
+    def enable(self, enable=True):
+        if self.is_enabled != enable:
+            self.element.click()
+
+    def disable(self):
+        self.enable(enable=False)
+
+
+class TopologyElement(object):
+    def __init__(self, obj, element):
+        if element is None:
+            raise KeyError('Element should not be None')
+        self.element = element
+        self.object = obj
+        element_data = re.search('Name: (.*) Type: (.*) Status: (.*)', element.text)
+        if len(element_data.groups()) != 3:
+            raise RuntimeError('Unexpected element')
```

Comment:
What unexpected element?

Reply:
I've just started rewriting those classes based on topology in web_ui. This is part from there. I'll take a closer look on that later.